### PR TITLE
Feature/salary calculator

### PR DESCRIPTION
**Pull request by ShiZinDle**
New feature - Salary Calculator:
allows users to select any category they previously created and create salary calculation settings for it. the feature also adds edit pages for the settings and view pages for the calculations. The calculation will be done for the chosen month and will be based on all events that are labeled with the corresponding category. Multiple salary settings are allowed, but only 1 for each user and category combination.

Further changes implemented:
-moved contents of file "app.database.database" to "app.database" for easier access.
-moved function "app.database,database.get_db" to "app.dependencies.get_db".
-merged file "tests.utils" into "app.internal.utils" to prevent code duplication.
-fixed imports and type hints where necessary.

### CODE REVIEW

**Comment by PureDreamer** [comment]
In order to pass the tests you need the files to be ok with flake8, you can choose flake8 as a linter, and it will show you where are your mistakes.

**Comment by ShiZinDle** [comment]
All linting errors, code conflicts and failing tests have been fixed and resolved.

File: app/routers/salary/utils.py [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+        date = end.date()
+    try:
+        return (datetime.combine(date, time(0)),
+                datetime.combine(date + timedelta(1),
+                                 time(0)))
+    except NameError:
+        return datetime.min, datetime.min
+
+
+def get_total_synchronous_hours(
```
Comment (yammesicka) [inline app/routers/salary/utils.py:132]: How about [this answer](https://stackoverflow.com/a/9044111/1058671)? :)

File: app/routers/salary/utils.py [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+        float: The weekly overtime amount for the supplied shifts.
+
+    Raises:
+        None
+    """
+    total_week_hours = sum(get_shift_len(shift.start, shift.end)
+                           for shift in shifts)
+    if total_week_hours <= wage.week_working_hours:
+        return 0.0
+    else:
```
Comment (yammesicka) [inline app/routers/salary/utils.py:278]: The `else:` is redundant here

File: app/routers/salary/utils.py [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+    if is_night_shift(start, end, wage):
+        hour_basis = wage.night_hour_basis
+    else:
+        hour_basis = wage.regular_hour_basis
+    shift_len = get_shift_len(start, end)
+
+    total_hours += get_hours_during_holiday(start, end, wage) / 2
+
+    if shift_len > hour_basis:
+        overtime += shift_len - hour_basis
```
Comment (yammesicka) [inline app/routers/salary/utils.py:222]: This should be a separate function :)

File: app/routers/salary/utils.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+
+    Args:
+        user_id (int): Id of the relevant user.
+        category_id (int): Id for the relevant category.
+
+    Returns:
+        SalarySettings | None: Settings for the provided user_id and
+                               category_id if exists, None otherwise.
+    """
+    session = SessionLocal()
```
Comment (yammesicka) [inline app/routers/salary/utils.py:470]: Prefer working with get_db dependency

**Review by yammesicka — CHANGES_REQUESTED** [review]
Wow! That's some effort you put in this PR. Awesome job!

File: app/routers/salary/config.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from datetime import time
```
[the remark this answers is not in the document]
Comment (ShiZinDle) [reply app/routers/salary/config.py:1]: Not sure how that happened, will resolve.

File: app/database/models.py [hunk 6]
```diff
@@ -110,6 +125,78 @@ def __repr__(self):
...
+class SalarySettings(Base):
+    # Code revision required after categories feature is added
+    # Code revision required after holiday times feature is added
+    # Code revision required after Shabbat times feature is added
+    __tablename__ = "salary_settings"
+
+    user_id = Column(
+        Integer, ForeignKey("users.id"), primary_key=True,
+    )
+    # category_id = Column(
```
[the remark this answers is not in the document]
Comment (yammesicka) [reply app/database/models.py:179]: Cool, marking as solved

File: app/internal/search.py [hunk 7]
```diff
@@ -16,7 +16,7 @@ def get_stripped_keywords(keywords: str) -> str:
 
 
 def get_results_by_keywords(
-                            session: SessionLocal,
+                            session: Session,
```
Comment (yammesicka) [inline app/internal/search.py:19]: Please use Depends(get_db) (also in other files, if applicable)

File: tests/salary/conftest.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    'create_settings': 'Already created your settings?',
+    'pick_settings': 'Edit Settings',
+    'edit_settings': 'Settings don\'t need editing?',
+    'pick_category': 'View Salary',
+    'view_salary': 'Need to alter your settings?',
+    'salary_calc': 'Total Salary:',
+}
+
+ROUTES = {
+    'home': '/salary',
```
Comment (yammesicka) [inline tests/salary/conftest.py:23]: Use url_for instead :)

**Review by yammesicka — CHANGES_REQUESTED** [review]
Awesome job, added few additional insights

File: app/internal/search.py (hunk 7 above)
Comment (ShiZinDle) [reply app/internal/search.py:19]: I don't think i fully understand your request here.
As I've mentioned before, you can see in [This comment by Tiangolo](https://github.com/tiangolo/fastapi/issues/1105#issuecomment-609919850), that Depends is not meant for use outside of path operation. Additionally, this function doesn't call a session directly, but is rather passed one from a path operation (where it is created using Depends(get_db)).
I should also mentioned that this is not part of my code and I merely fixed the incorrect type hinting in this module.

Comment (yammesicka) [reply app/internal/search.py:19]: I see, you are right. Sababa, let's keep it this way :)

**Review by yammesicka — CHANGES_REQUESTED** [review]
Awesome, a little fix and it's ready to merge from my side. Please also solve the conflicts :)

File: app/routers/salary/utils.py [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+        category_id (int): Id for the relevant category.
+
+    Returns:
+        SalarySettings | None: Settings for the provided user_id and
+                               category_id if exists, None otherwise.
+    """
+    settings = session.query(SalarySettings).filter(and_(
+        SalarySettings.user_id == user_id, SalarySettings.category_id
+        == category_id)).first()
+    session.close()
```
Comment (yammesicka) [inline app/routers/salary/utils.py:488]: Is this necessary?

Comment (ShiZinDle) [reply app/routers/salary/utils.py:488]: I can't really say why, but for some reason removing this line fails one of my tests.
The functionality is essentially the same, and the service itself doesn't seem to be affected, yet I didn't manage to pass the tests without it.
If you want to take a look at it, the one that fails is "test_routes: :test_edit_settings".

Comment (yammesicka) [reply app/routers/salary/utils.py:488]: Try to `git checkout upstream/develop -- requirements.txt` and `pip install -U -r requirements.txt`.
We've downgraded the pytest-asyncio library, that's might be the problem

Comment (ShiZinDle) [reply app/routers/salary/utils.py:486]: Unfortunately this did not solve the problem.

**Review by yammesicka — APPROVED** [review]

**Comment by yammesicka** [comment]
While trying to solve conflicts, I've seen there is a minor problem with test_events.py. Can you please take a look? :)

**Comment by ShiZinDle** [comment]
> While trying to solve conflicts, I've seen there is a minor problem with test_events.py. Can you please take a look? :)

There was a missing import, now fixed.

**Pull request merged**